mlir/lib/Target/LLVM/ROCDL/Utils.cpp

```diff
@@ -0,0 +1,230 @@
+#include "llvm/Support/AMDGPUMetadata.h"
+
+using namespace mlir;
+using namespace mlir::ROCDL;
+
+/// Search the ELF object and return an object containing the `amdhsa.kernels`
+/// metadata note. Function adapted from:
+/// llvm-project/llvm/tools/llvm-readobj/ELFDumper.cpp Also see
+/// `amdhsa.kernels`:
+/// https://llvm.org/docs/AMDGPUUsage.html#code-object-v3-metadata
+template <typename ELFT>
+static std::unique_ptr<llvm::msgpack::Document>
+getAMDHSANote(llvm::object::ELFObjectFile<ELFT> &elfObj) {
+  using namespace llvm;
+  using namespace llvm::object;
+  using namespace llvm::ELF;
+  const ELFFile<ELFT> &elf = elfObj.getELFFile();
+  Expected<typename ELFT::ShdrRange> secOrErr = elf.sections();
+  if (!secOrErr) {
+    consumeError(secOrErr.takeError());
+    return nullptr;
+  }
+  ArrayRef<typename ELFT::Shdr> sections = *secOrErr;
+  for (const typename ELFT::Shdr &section : sections) {
+    if (section.sh_type != ELF::SHT_NOTE)
+      continue;
+    size_t align = std::max(static_cast<unsigned>(section.sh_addralign), 4u);
+    Error err = Error::success();
+    for (const typename ELFT::Note note : elf.notes(section, err)) {
+      StringRef name = note.getName();
+      if (name != "AMDGPU")
+        continue;
+      uint32_t type = note.getType();
+      if (type != ELF::NT_AMDGPU_METADATA)
+        continue;
+      ArrayRef<uint8_t> desc = note.getDesc(align);
+      StringRef msgPackString =
+          StringRef(reinterpret_cast<const char *>(desc.data()), desc.size());
+      std::unique_ptr<llvm::msgpack::Document> msgPackDoc(
+          new llvm::msgpack::Document());
+      if (!msgPackDoc->readFromBlob(msgPackString, /*Multi=*/false))
+        return nullptr;
+      if (msgPackDoc->getRoot().isScalar())
+        return nullptr;
+      return msgPackDoc;
+    }
+  }
+  return nullptr;
+}
+
+/// Return the `amdhsa.kernels` metadata in the ELF object or nullptr on
+/// failure. This is a helper function that casts a generic `ObjectFile` to the
+/// appropiate `ELFObjectFile`.
+static std::unique_ptr<llvm::msgpack::Document>
+getAMDHSANote(ArrayRef<char> elfData) {
+  using namespace llvm;
+  using namespace llvm::object;
+  if (elfData.empty())
+    return nullptr;
+  MemoryBufferRef buffer(StringRef(elfData.data(), elfData.size()), "buffer");
+  Expected<std::unique_ptr<ObjectFile>> objOrErr =
+      ObjectFile::createELFObjectFile(buffer);
+  if (!objOrErr || !objOrErr.get()) {
+    // Drop the error.
+    llvm::consumeError(objOrErr.takeError());
+    return nullptr;
+  }
+  ObjectFile &elf = *(objOrErr.get());
+  if (auto *obj = dyn_cast<ELF32LEObjectFile>(&elf))
+    return getAMDHSANote(*obj);
+  else if (auto *obj = dyn_cast<ELF32BEObjectFile>(&elf))
+    return getAMDHSANote(*obj);
+  else if (auto *obj = dyn_cast<ELF64LEObjectFile>(&elf))
+    return getAMDHSANote(*obj);
+  else if (auto *obj = dyn_cast<ELF64BEObjectFile>(&elf))
+    return getAMDHSANote(*obj);
+  return nullptr;
+}
```

Comment:
@jhuber6 , @shiltian could you take a look at this code for extracting AMDGPU object metadata from ELF? I think eventually this should be migrated to Offload (maybe you already have it and I didn't notice it).

Reply:
Thank you for the hint of where to find it, eventually I think we (MLIR) should move to use those offload utilities and consolidate.
Good pointer on `ELF64LE`. Besides that, any other comments?

Reply:
I don't know if we're in a position where `mlir` can depend on `offload` here, possibly we could just move that to some LLVM utility. I don't know much of anything about MLIR unfortunately.

Reply:
Yeah, I  100% agree. Now, long term it would be ideal to consolidate many of these somewhere. I think if we are able to make offload modular enough (wrt the build process), it shouldn't be too much of a problem.